include/algorithms/public/KNNRegressor.hpp

```diff
@@ -62,13 +63,17 @@ class KNNRegressor
       }
     }
 
+    RealVector prediction(targets.pointSize(),alloc); //should we make a private allocation once? or just write in prediction directly
```

Comment:
I do to accumulate in it, no?

Reply:
If we have an `output` parameter, you can accumulate straight into that. If we have a return value, we can accumulate into that instead. Horses for courses. In any case, my preference is to get rid of the hand-rolled accumulation.

Reply:
but you assign a matrix of k by dim to do so, isn't that just more abstract and costs more memory?

Reply:
No I don't, and if I did, it would only matter secondarily to maintainability (as in, don't fix hypothetical performance problems before they're measurably problematic, especially if it makes for gnarlier code). 

What I do is wrap the target dataset's data in an `Eigen::Map` (no allocation) and return a sub view of that with just the rows we want. 

Because I wrote 
```c++ 
  auto targetPoints = asEigen<Array>(targets.getData())(indices, Eigen::all);
```

as far as Eigen is concerned this is just an *expression*, waiting to be lazily evaluated, rather than an object. On the other hand, if I'd writtem 

```c++ 
Eigen::ArrayXXd  targetPoints = asEigen<Array>(targets.getData())(indices, Eigen::all);
```

then that would have triggered the allocation and copy (because `Eigen::ArrayXXd` is explicitly making a new container). I.e., what's important there is using `auto` to keep the thing as an Eigen expression.